ast/visitor.go

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2021 ByteDance Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ast
+
+import (
+    `encoding/json`
+
+    `github.com/bytedance/sonic/internal/native/types`
+)
+
+// Visitor handles the callbacks during preorder traversal of a JSON AST.
+//
+// According to the JSON RFC8259, a JSON AST can be defined by
+// the following rules without seperator / whitespace tokens.
+//
+//  JSON-AST  = value
+//  value     = false / null / true / object / array / number / string
+//  object    = begin-object [ member *( member ) ] end-object
+//  member    = string value
+//  array     = begin-array [ value *( value ) ] end-array
+//
+type Visitor interface {
+
+    // OnNull handles a JSON null value.
+    OnNull() error
+
+    // OnTrue handles a JSON true value.
+    OnTrue() error
```

Comment:
could we use OnBool(v bool) error? likes https://json.nlohmann.me/features/parsing/sax_interface/